### PR TITLE
Add function for converting a tree to its prufer sequence

### PR DESCRIPTION
**Pull request by frederik-h**
This is an implementation of the feature requested in #1084.

Regarding licensing: I'm fine with GPL 2 or later and FDL.

### CODE REVIEW

File: src/conversion.c [hunk 1]
```diff
@@ -844,3 +844,93 @@ int igraph_get_stochastic_sparsemat(const igraph_t *graph,
...
+     sequence. If the removal of u turns the neighbor into a leaf which is < u,
+     we repeat the procedure for the new leaf and so on. */
+  igraph_integer_t u;
+  igraph_vector_t degrees, neighbors;
+  igraph_integer_t prufer_index = 0;
+  igraph_integer_t n = igraph_vcount(graph);
+
+  if(n < 2)
+    return IGRAPH_EINVAL;
+
```
Comment (frederik-h) [inline src/conversion.c]: Should we also check that the graph is a tree?

Comment (szhorvat) [reply src/conversion.c]: I think that would be useful.

File: src/conversion.c [hunk 2]
```diff
@@ -844,3 +844,93 @@ int igraph_get_stochastic_sparsemat(const igraph_t *graph,
...
+     If u is a leaf, we remove it and add its unique neighbor to the prüfer
+     sequence. If the removal of u turns the neighbor into a leaf which is < u,
+     we repeat the procedure for the new leaf and so on. */
+  igraph_integer_t u;
+  igraph_vector_t degrees, neighbors;
+  igraph_integer_t prufer_index = 0;
+  igraph_integer_t n = igraph_vcount(graph);
+
+  if(n < 2)
+    return IGRAPH_EINVAL;
```
Comment (szhorvat) [inline src/conversion.c]: On error, do not simply return, use `IGRAPH_ERROR` instead and report a user-readable error. `IGRAPH_ERROR` also ensures that the error handler will be called and any allocated resources can be freed (should a future version of this function allocate resources before this line).

File: src/conversion.c [hunk 3]
```diff
@@ -844,3 +844,93 @@ int igraph_get_stochastic_sparsemat(const igraph_t *graph,
...
+
+  IGRAPH_CHECK(igraph_vector_int_resize(prufer, n-2));
+  IGRAPH_CHECK(igraph_vector_init(&degrees, n));
+  IGRAPH_CHECK(igraph_vector_init(&neighbors, 1));
+
+  IGRAPH_FINALLY(&degrees, &igraph_vector_destroy);
```
Comment (szhorvat) [inline src/conversion.c]: The `IGRAPH_FINALLY` should follow the `igraph_vector_init` that it is paired with immediately. Consider that the *next* `igraph_vector_init` may fail too.

You can consider using `IGRAPH_VECTOR_INIT_FINALLY` to do both initialization and `IGRAPH_FINALLY` in a single line.

File: src/conversion.c [hunk 4]
```diff
@@ -844,3 +844,93 @@ int igraph_get_stochastic_sparsemat(const igraph_t *graph,
...
+    while(degree == 1 && leaf <= u) {
+      igraph_integer_t i;
+      igraph_integer_t neighbor = 0;
+
+      VECTOR(degrees)[leaf] = 0; /* mark leaf v as deleted */
+
+      IGRAPH_CHECK(igraph_neighbors(graph, &neighbors, leaf, IGRAPH_ALL));
+
+      /* Find the unique remaining neighbor of the leaf */
+      for(i = 0; i < igraph_vector_size(&neighbors); i++) {
```
Comment (szhorvat) [inline src/conversion.c]: I would call `igraph_vector_size` *once* before the loop and store the vector size (as function calls can be slow when inlining is prevented, like in this case).

**Comment by frederik-h** [comment]
> I am not a maintainer of igraph, just trying to help with the review. 
> I do suggest adding a check for whether the input graph is a tree. (...)
> I did test the code a bit by integrating it into the Mathematica interface.

Thanks! I have added the check.

**Comment by szhorvat** [comment]
Thank you for the contribution. I wanted to do this for a while myself, but I just couldn't get around to it.

**Comment by ntamas** [comment]
Thanks for the contribution guys. I'm on holiday in Cyprus till the end of the week, I'll try to get around to reviewing this next week when I am back.

**Comment by gaborcsardi** [comment]
Thanks both!

**Comment by Pascal-Ortiz** [comment]
> This is an implementation of the feature requested in #1084.
> 
Hi Frederik!

Good work! there are not so many libraries implementing Prufer conversions. 

Looking at the source code, I thought your algorithm has linear complexity. To clarify, I have done some testing and it appears that execution seems to be quadratic (if you double the data size then time execution is multiplied by a factor 4) and the implentation was not able to build in a reasonable time a tree having one million of nodes from a prufer sequence.

**Comment by szhorvat** [comment]
@Pascal-Ortiz 

> and the implentation was not able to build in a reasonable time a tree having one million of nodes from a prufer sequence.

This PR was for a function that takes a tree and creates a Prüfer sequence. You seem to be talking about taking a Prüfer sequence and building a tree. Can you clarify which one you are referring to?

**Comment by Pascal-Ortiz** [comment]
I'm talking about the `igraph_from_prufer` function.

**Comment by Pascal-Ortiz** [comment]
> This PR was for a function that takes a tree and creates a Prüfer sequence.

Sorry for the confusion. PR #1079 is the appropriate one.

**Comment by szhorvat** [comment]
@Pascal-Ortiz Can you please open a new issue and describe the problem in detail, along with a link to the linear-time one? (So that I wouldn't forget about this—I did a quick test and indeed it's much slower than it should be.)

**Comment by szhorvat** [comment]
Also a note:

If you want to use this for generating random trees, there's a different function for that (`igraph_tree_game`) which has a method that will be (much) faster for large trees.

**Pull request merged**